src/core/server/server.cc

```diff
@@ -84,6 +86,227 @@
+}
+
+//
+// Server::ListenerInterface
+//
+
+void Server::ListenerWrapper::Start() {
+  if (IsServerListenerEnabled()) {
+    if (server_->config_fetcher() != nullptr) {
+      auto watcher =
+          std::make_unique<ListenerWrapper::ConfigFetcherWatcher>(this);
+      config_fetcher_watcher_ = watcher.get();
+      server_->config_fetcher()->StartWatch(
+          grpc_sockaddr_to_string(listener_->resolved_address(), false).value(),
+          std::move(watcher));
+    } else {
+      {
+        MutexLock lock(&mu_);
+        started_ = true;
+        is_serving_ = true;
+      }
+      listener_->Start();
+    }
+  } else {
+    listener_->Start();
+  }
+}
+
+void Server::ListenerWrapper::Stop() {
+  if (IsServerListenerEnabled()) {
+    absl::flat_hash_set<OrphanablePtr<ListenerInterface::LogicalConnection>>
+        connections;
+    {
+      MutexLock lock(&mu_);
+      // Orphan the connections so that they can start cleaning up.
+      connections = std::move(connections_);
+      is_serving_ = false;
+    }
+    if (config_fetcher_watcher_ != nullptr) {
+      CHECK_NE(server_->config_fetcher(), nullptr);
+      server_->config_fetcher()->CancelWatch(config_fetcher_watcher_);
+    }
+  }
+  GRPC_CLOSURE_INIT(&destroy_done_, ListenerDestroyDone, server_,
+                    grpc_schedule_on_exec_ctx);
+  listener_->SetOnDestroyDone(&destroy_done_);
+  listener_.reset();
+}
+
+absl::optional<ChannelArgs> Server::ListenerWrapper::AddLogicalConnection(
+    OrphanablePtr<ListenerInterface::LogicalConnection> connection,
+    const ChannelArgs& args, grpc_endpoint* endpoint) {
+  RefCountedPtr<grpc_server_config_fetcher::ConnectionManager>
+      connection_manager;
+  {
+    MutexLock lock(&mu_);
+    if (!is_serving_) {
+      // Not serving
+      return absl::nullopt;
+    }
+    connection_manager = connection_manager_;
+  }
+  ChannelArgs new_args = args;
+  if (server_->config_fetcher() != nullptr) {
+    if (connection_manager == nullptr) {
+      // Connection manager not available
+      return absl::nullopt;
+    }
+    absl::StatusOr<ChannelArgs> args_result =
+        connection_manager->UpdateChannelArgsForConnection(new_args, endpoint);
+    if (!args_result.ok()) {
+      return absl::nullopt;
+    }
+    auto* server_credentials =
+        (*args_result).GetObject<grpc_server_credentials>();
+    if (server_credentials == nullptr) {
+      // Could not find server credentials
+      return absl::nullopt;
+    }
+    auto security_connector =
+        server_credentials->create_security_connector(*args_result);
+    if (security_connector == nullptr) {
+      // Unable to create secure server with credentials
+      return absl::nullopt;
+    }
+    new_args = (*args_result).SetObject(security_connector);
+  }
+  MutexLock lock(&mu_);
+  if (!is_serving_ || connection_manager != connection_manager_) {
+    // Not serving
+    return absl::nullopt;
+  }
+  connections_.emplace(std::move(connection));
+  return new_args;
```

Comment:
Instead of having this method return the new args, how about adding a method to `LogicalConnection` like this:

```
virtual void Start(const ChannelArgs& args) = 0;
```

Then we can call that here.

Reply:
I would say that the goal of the `ListenerState` interface is actually to provide whatever interface the server needs to expose to the transport-specific listener and connection code; I think the config management and connection tracking functionality is definitely part of that, but only inasmuch as those things are part of the functionality provided by the server.  In that context, I think starting the `LogicalConnection` could be considered part of adding the `LogicalConnection` to the server, so it wouldn't bother me to call `Start()` here.

That having been said, I don't think this matters very much one way or the other, so it's fine if you prefer to leave it as-is.